Add sbert implementation for inference

lerouxrgd opened the pull request:
This PR adds a generic implementation for [Sentence-Transformers](https://sbert.net/) inference #185.

The main components are:
* The `SBertModel` struct templated by `T: SBertTransformer` (which is a [sealed trait](https://rust-lang.github.io/api-guidelines/future-proofing.html))
* Layers are defined based on `modules.json` (support for pooler and optional dense/normalization)
* Implementations of `SBertTransformer`: 
  * `UsingAlbert`, `UsingBert` , `UsingDistilBert`, `UsingRoberta`, `UsingT5`
  * All tested against results from Python implementations available on [Hugging Face Hub](https://huggingface.co/sentence-transformers)
  * Tests are marked with `#[ignore]` as they require to download and convert the models beforehand (instructions are in the tests comments)
* The script `utils/convert_model.py` has new parameters to be able to convert sbert models to Rust.

This PR is marked as draft because:
* Currently `UsingT5` is not working as `T5Model` always define a decoder and sbert models contain only an encoder. I don't know what would be the best approach for "encoder only T5 model", do you have any preference/insight ?
* Maybe there are more implementations to add ? If there are more models to support I can add them.
* Maybe `SBertTransformer` doesn't need to be sealed, what do you think ?
* Attention in `UsingAlbert` is averaged across the hidden groups to have the type `Vec<Tensor>` (instead of `Vec<Vec<Tensor>>`), which is the type of all other implementations. Is this OK for you ?

guillaume-be commented:
Hello @lerouxrgd ,

Thank you for the first draft on this implementation. I have started to give the implementation some thoughts - and while I haven't had the time to implement I started to make an implementation plan for this feature. I would like the sentence embeddings to be aligned with the rest of the pipelines in this crate from an implementation point of view (e.g. `ModelForQuestionAnswering` or `ModelForSequenceClassification`). 

Please see below a few comments that hopefully will make sense (these are rather big changes so I will not comment directly on the files at this point):
- Can you create a new pipeline module (`sentence_embeddings.rs` or a subfolder with multiple files if needed as done for `translation` under `/pipeline`) containing the new functionality implemented as part of this PR.
- Create a new `SentenceEmbeddingsConfig` (stored in the sentence embeddings module in `/pipelines`) that would be created from the configuration specific to sentence embeddings `modules.json`)
- Move the definition of common utility layers/modules for sentence embeddings to the pipeline module (hence possibly the need for multiple files)
- Create new task-specific heads for the models (e.g. create a `BertForSentenceEmbeddings` similar to how `BertForQuestionAnswering` is defined) that would take **two** configurations (the model-specific configuration and the sentence emebddings configuration defined above). Implement the same for all models commonly used for sentence embeddings. The benefit of this approach is to allow for model-specific implementations of sentence embeddings. As you have mentioned, T5 for sentence embeddings just uses an encoder: with this model the `T5ForSentenceEmbeddings` can be created this way.
- Create Enum with accepted variants, aligned with how it is done for other pipelines wrapping the new task-specific models created.

I believe this implementation would follow the general structure of the crate a bit better - but very happy to discuss/refine further!

lerouxrgd commented:
Thank you for the feedback ! The big picture is clear, I'll try to adapt the PR accordingly. I might have questions while doing so.

lerouxrgd commented:
@guillaume-be I have refactored the sentence embeddings logic into a dedicated pipeline.

I was able to customize T5 for this and now it works fine. My only concern for T5 is that it's not easy to configure whether to output attentions or not (it's the only remaining TODO). I think the simplest way to handle it would be to have a field `output_attentions: Option<bool>` in the `T5Config`, what do you think ?

Let me know if you have any other remarks.

guillaume-be commented:
> @guillaume-be I have refactored the sentence embeddings logic into a dedicated pipeline.
> 
> I was able to customize T5 for this and now it works fine. My only concern for T5 is that it's not easy to configure whether to output attentions or not (it's the only remaining TODO). I think the simplest way to handle it would be to have a field `output_attentions: Option<bool>` in the `T5Config`, what do you think ?
> 
> Let me know if you have any other remarks.

Thank you for working on this again! I will have a closer look over the weekend and get back to you in a few days. For T5 you are right - for some reason the `output_attentions` and `output_hidden_states` are missing from the configuration - can you please add those within this PR?

guillaume-be commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+        }
+    }
+
+    pub fn encode_with_attention<S>(
```
Is there a use case to justify the added complexity? As far as I know the reference `sentence-transformers` only has an `encode` method: https://www.sbert.net/docs/package_reference/SentenceTransformer.html

lerouxrgd replied:
Right, similarly to my other comment, this has been helpful for explainability/vizualisation features in our use-cases.
I think that being able to get attentions one way or another is important with this kind of models.

guillaume-be commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+pub struct SentenceEmbeddingsTokenizerOuput {
```
This probably does not need to be public

lerouxrgd replied:
Sure, it needs to be public only if `tokenize(...)` remains public too.

guillaume-be commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+
+pub struct SentenceEmbeddingsModelOuput {
+    pub embeddings: Tensor,
+    pub all_attentions: Option<Vec<Tensor>>,
```
I would suggest simplifying and either directly returning a Tensor or have `SentenceEmbeddingsModelOuput ` containing a reference to the input sentence and the embeddings, i.e.:
```rust
SentenceEmbeddingsModelOuput {
    pub sentence: &str,
    pub embeddings: Tensor,
}
```

lerouxrgd replied:
Trying to pass a reference to the original sentence seems tricky here, I'd rather have just a `Tensor` if you want the attentions removed.

guillaume-be commented on src/pipelines/sentence_embeddings/mod.rs:
```diff
@@ -0,0 +1,21 @@
+    SentenceEmbeddingsTokenizerOuput,
+};
+
+pub type Attention = Vec<f32>; // Length = sequence length
```
If the attention output is not included in the Sentence Embeddings output (see other comments), these types could be removed.

lerouxrgd replied:
Sure, let me know what you decide.

guillaume-be commented on src/t5/t5_model.rs:
```diff
@@ -133,6 +133,8 @@ pub struct T5Config {
     pub feed_forward_proj: Option<FeedForwardProj>,
     pub tie_word_embeddings: Option<bool>,
     task_specific_params: Option<TaskSpecificParams>,
+    pub output_attentions: Option<bool>,
```
Regardless of the decision on the attentions inclusion for the sentence emebeddings, please keep these changes, and thank you again for catching this.

guillaume-be commented on src/pipelines/sentence_embeddings/layers.rs:
```diff
@@ -0,0 +1,134 @@
+pub struct Dense {
+    linear: nn::Linear,
+    activation: TensorFunction,
+    _var_store: nn::VarStore,
```
Are `_var_store` and `_conf` used at a later point?

lerouxrgd replied:
Regarding `_var_store`, since some weights are loaded from another file specifically for this layer (in the original Python implementation), I think it's good to keep it as part of the struct in order to have a clean lifecycle (i.e. the layer is dropped => this sub VarStore is dropped too).

As for `_conf` I kept it in case it would be used in `forward(...)`, or for debugging purposes. But I can remove it if you prefer.

guillaume-be replied:
I see I understand now - one alternative would be to manage the `var_store` from the SentenceEmbeddingsModel`, as is done for the transformer. This is also fine. I would however remove the `_config` object, it's value can be observed from `SentenceEmbeddingsModel` when debugging anyway.

lerouxrgd replied:
Done

guillaume-be reviewed (commented):
Thank you very much @lerouxrgd for making the change s- this looks great!
I have added some comments directly in the code - mainly related to questions about the inclusion of the attentions in the output of the sentence embeddings which seems to add a non-negligible amount of complexity to the code (while the reference Python implementation does not include this option).

Please have a look and let me know what you think.
Thanks again --

lerouxrgd commented on src/common/activations.rs:
```diff
@@ -58,6 +62,8 @@ pub enum Activation {
     /// Tanh
     tanh,
+    /// Identity
+    identity,
```
It's used in `sentence-t5-base/2_Dense/config.json`:

```json
{"in_features": 768, "out_features": 768, "bias": false, "activation_function": "torch.nn.modules.linear.Identity"}
```

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+            _ => Some(false),
+        };
+        let add_prefix_space = None;
+        let tokenizer = TokenizerOption::from_file(
```
In `token_classification.rs` the lines just above your examples are:

```rs
        let config_path = config.config_resource.get_local_path()?;
        let vocab_path = config.vocab_resource.get_local_path()?;
        let weights_path = config.model_resource.get_local_path()?;
```

which is the same as what is here, but with intermediate variables. I can add intermediate variables if it's better. 

Or is it something else that you mean by a bit lighter ?

lerouxrgd replied:
Besides, I can move `strip_accents` and `add_prefix_space` to `SentenceEmbeddingsConfig` if you prefer.

However their default values (`None` or `Some(false)`) can be tricky to define beforehand (I mean you need to know the model for that).

guillaume-be replied:
Thank you - I was referring to the transpose, multiple maps and conversion to an owned string that needs to be converted to a reference (as opposed to `vocab_path.to_str().unwrap()` - but this is also fine - feel free to leave as is.

Yes - could you please move the `strip_accents` and `add_prefix_space` to the `SentenceEmbeddingsTokenizerConfig` together with `lower_case` instead? It seems that is the place where this is defined (see https://huggingface.co/sentence-transformers/msmarco-roberta-base-v3/blob/main/tokenizer_config.json). The `TokenizerOption` instantiation should take care of defining the default value if not provided already - there should be no need to define the default value in this pipeline directly.

lerouxrgd replied:
Done

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+        self.tokenizer_truncation_strategy = truncation_strategy;
+    }
+
+    pub fn tokenize<S>(&self, inputs: &[S]) -> SentenceEmbeddingsTokenizerOuput
```
From my experience with debugging those models, being able to inspect tokenization results greatly helps, which is why I put it public.

lerouxrgd replied:
Furthermore it's present in the [Python implementation](https://www.sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.tokenize)

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+
+        Ok(SentenceEmbeddingsModelOuput {
+            embeddings: maybe_normalized,
+            all_attentions,
```
In our use-cases having the attention helps for defining some kind of explainability (which heads looks at what). There are also some visualization methods based on this: https://github.com/jessevig/bertviz

guillaume-be replied:
Ok -

lerouxrgd commented on src/bert/bert_model.rs:
```diff
@@ -1179,6 +1179,41 @@ impl BertForQuestionAnswering {
     }
 }
 
+pub struct BertForSentenceEmbeddings {
```
Sure, it can be done independently from the attention anyway.

lerouxrgd replied:
Done

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+        }
+    }
+
+    pub fn forward<S>(&self, inputs: &[S]) -> Result<SentenceEmbeddingsModelOuput, RustBertError>
```
I feel it's interesting to be able to have it in case you want the result as a `Tensor` (all `forward(...)` methods return `Tensor`s) and let `encode(...)` be specialized by returning a `Vec<Embedding>`.

guillaume-be replied:
I understand - can we instead make `forward` private, and create a new `encode_as_tensors` method that would return the tensor output? My concern is the difference between forward and encode not being clear, and I believe it would be easier if all methods would be called `encode_[...]` describing what they return.

guillaume-be replied:
Could you please also make sure that all public structures and methods have a docstring, taking as a model the other pipelines?

lerouxrgd replied:
> I understand - can we instead make `forward` private, and create a new `encode_as_tensors` method that would return the tensor output? My concern is the difference between forward and encode not being clear, and I believe it would be easier if all methods would be called `encode_[...]` describing what they return.

Done

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+    where
+        S: AsRef<str> + Sync,
+    {
+        let SentenceEmbeddingsModelOuput { embeddings, .. } = self.forward(inputs)?;
```
If `output_attentions` is set to `false` in the config they are `None` anyway.

lerouxrgd replied:
Having the attention (if kept in the PR) returned here is consistent with the other `forward(...)` methods.

lerouxrgd commented on src/pipelines/sentence_embeddings/config.rs:
```diff
@@ -0,0 +1,187 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SentenceEmbeddingsModules(pub Vec<SentenceEmbeddingsModule>);
```
Done

lerouxrgd commented on src/pipelines/sentence_embeddings/config.rs:
```diff
@@ -0,0 +1,187 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SentenceEmbeddingsModule {
```
Done

lerouxrgd commented on src/pipelines/sentence_embeddings/pipeline.rs:
```diff
@@ -0,0 +1,370 @@
+        // Setup dense layer
+
+        let mut dense_layer = None;
+        if modules.dense_module().is_some() {
```
Done

lerouxrgd commented:
@guillaume-be  I think I have addressed all the points you mentioned, expect the doc strings. I'll add them in another commit. Let me know if I have forgotten something.

lerouxrgd commented:
@guillaume-be is there any updates regarding the status of this PR ? Or are there things that you want added/changed ?

lerouxrgd commented:
Awesome, thanks for the heads up !
Your addition of remote resources looks pretty useful !
I think we can even simplify their usage with a `SentenceEmbeddingsBuilder<Remote>`, I will add a commit with it soon.

guillaume-be commented:
Thank you very much for this great PR @lerouxrgd !

Fixes https://github.com/guillaume-be/rust-bert/issues/168, https://github.com/guillaume-be/rust-bert/issues/185
